Fix deadlocks while running test

gmarcos87 opened the pull request:
This pull request resolves:
- [x] removes "...DatabaseError: deadlock detected.." https://github.com/opencollective/opencollective/issues/2623

znarf commented on .circleci/config.yml:
```diff
@@ -71,6 +71,7 @@ jobs:
       - run: mkdir -p $CIRCLE_ARTIFACTS $CIRCLE_TEST_REPORTS
       - run: npm ci
       - run: npm run db:setup
+      - run: npm run db:migrate
```
What is the impact of the migration? It was skipped on purpose, the model is supposedly already up to date, isn't it?

gmarcos87 replied:
Without migration sequelize.sync does not find the tables in postgres and displays the error before creating each one. That's why running the migrations avoids the error log in the tests.

znarf replied:
Thank you, I'll dig deeper into it.

znarf replied:
So, after investigation, the outcome is great, we do want to have the schema created before running the tests.

However, I don't think it's good idea and create the schema with `npm run db:migrate`, it's slow and ineffective. I think it's better to use `sequelize.sync` and add it in the setup script.

I explored this solution in: https://github.com/opencollective/opencollective-api/pull/2973

znarf commented on test/graphql.paymentMethods.test.js:
```diff
@@ -15,7 +15,11 @@ import { randEmail } from './stores';
-  beforeEach(() => utils.resetTestDB());
+  beforeEach(async () => {
+    await utils.resetTestDB();
+    await utils.resetTestModelDb('CollectiveHistory');
```
Why do we need to to that? What does that solve?

gmarcos87 replied:
After many tests I came to the conclusion that it can be changed. Its original goal was to ensure the creation of the table "CollectiveHistory" which failed repeatedly.
I think the problem is due to an error in sequelize: sequelize.sync not completed when resolving (https://github.com/sequelize/sequelize/issues/8918) or by asynchronous events from which no resolution is expected (I saw some in the hooks) and that events/functions use the database that is being recreated by sequelize.sync at the same time.

If I add time before and after running sequelize.sync all problems disappear. The amount of time varies depending on the execution environment (memory, processor, Internet connection, etc.).

During the debugging process the only thing I noticed that improved the success rate was to enable the __logQueryParameters__ and __logging__ options. 😯 I suppose it slows down some processes and thereby giving others time to conclude gracefully.

In the case of the file ```graphql.paymentMethods.test``` (where the problem with "CollectiveHistory" occurs) I had to add time before running sequelize.sync, which gives me to understand that some postgres or sequelize process is announced as finished when it actually continues to block writing, or that some hook continues to work asynchronously while table deletion occurs.
_________________
Rather than worrying about the appearance of logs, what I would like to understand is the impact of these lags on production. I don't think there is any creation/droppingse of tables in production, so this is not a threat, but it can be remembered as something to revisit.

gmarcos87 commented:
Great, I can do a rebase from ##2973, include that commit and remove the ones that aren't needed.

znarf commented on test/graphql.mutation.test.js:
```diff
@@ -49,7 +49,11 @@ describe('Mutation Tests', () => {
 
-  beforeEach('reset db', () => utils.resetTestDB());
+  beforeEach('reset db', async () => {
+    await new Promise(res => setTimeout(res, 500));
```
Could we replace that by a try/catch/wait/retry? I'm afaid to slow down all tests by 500ms.

gmarcos87 replied:
try/catch do not fulfill the same function. The problem is that there are actions that continue to need the db even after the test case is over. This 500 millisecond wait gives them time to finish before the db is deleted and recreated. It doesn't happen all the time just in a couple of cases (it's only in graphql.mutation.test.js), the total time increases only in two or three seconds.

znarf replied:
Ok, hadn't noticed it was only it's only in graphql.mutation.test.js.

I was thinking about.

```
try {
  await utils.resetTestDB()
} catch (e) {
  await new Promise(res => setTimeout(res, 500));
  await utils.resetTestDB()
}
```

gmarcos87 replied:
utils.resetTestDB() doesn't fail, what fails are the processes that were left in the background and over which we can't catch the errors.

znarf commented:
Looks like the rebase was not properly done, my changes are appearing in the PR.

gmarcos87 commented:
> Looks like the rebase was not properly done, my changes are appearing in the PR.

I had rebased to #2973. I removed the commit and updated it with master.
